[TST] Test cimporting Cython APIs

thalassemia opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
Closes gh-16748.

#### What does this implement/fix?
Test the Cython APIs for BLAS, LAPACK, optimize, and special in separately compiled Cython modules. 

#### Additional information
I'm not sure how to best consolidate the duplicated code.

j-bowhay commented on scipy/linalg/tests/test_extending.py:
```diff
@@ -0,0 +1,79 @@
+except ImportError:
+    cython = None
+else:
+    from numpy._utils import _pep440
```
It might be better to use our own version of `pep440` (`from scipy._lib import _pep440`) rather than relying on a private api

lucascolley commented on scipy/linalg/tests/test_extending.py:
```diff
@@ -0,0 +1,79 @@
+    c_cy = np.ones(8) * 4
+    x_cy = np.ones(9)
+    extensions.tridiag(a_cy, b_cy, c_cy, x_cy)
+    assert np.array_equal(x, x_cy)
```
Using `np.testing.assert_array_equal` is standard for SciPy.

thalassemia committed:
TST: Test cimporting Cython APIs

thalassemia commented:
This PR is already paying off. Looks like `cython_special.pxd` was accidentally dropped from the final install amidst the bigger (and very exciting!) updates in gh-20417. It was causing CI failures here so I added it as a commit. Should I make it a separate PR instead?

Took me a couple attempts to figure out the MyPy errors. I added comments about how my type-check skips will need to be updated when we start using MyPy 1.6+ (currently 1.0.0).

j-bowhay commented on scipy/linalg/tests/test_extending.py:
```diff
@@ -0,0 +1,81 @@
+@pytest.mark.skipif(cython is None, reason="requires cython")
+@pytest.mark.slow
+def test_cython(tmp_path):
+    # build the examples in a temporary directory
+    srcdir = os.path.join(os.path.dirname(__file__), '..')
+    shutil.copytree(srcdir, tmp_path / 'linalg')
+    build_dir = tmp_path / 'linalg' / 'tests' / '_cython_examples'
+    target_dir = build_dir / 'build'
+    os.makedirs(target_dir, exist_ok=True)
+    if sys.platform == "win32":
+        subprocess.check_call(["meson", "setup",
+                               "--buildtype=release", 
+                               "--vsenv", str(build_dir)],
+                              cwd=target_dir,
+                              )
+    else:
+        subprocess.check_call(["meson", "setup", str(build_dir)],
+                              cwd=target_dir
+                              )
+    subprocess.check_call(["meson", "compile", "-vv"], cwd=target_dir)
+
+    # import without adding the directory to sys.path
+    suffix = sysconfig.get_config_var('EXT_SUFFIX')
+
+    def load(modname):
+        so = (target_dir / modname).with_suffix(suffix)
+        spec = spec_from_file_location(modname, so)
+        mod = module_from_spec(spec)
+        spec.loader.exec_module(mod)
+        return mod
+
+    # test that the module can be imported
+    extensions = load("extending")
```
This logic could presumably be put in a function in `_lib`?

thalassemia replied:
Great idea! I refactored it into a function in `scipy._lib._testutils`. This resolves most of my concerns regarding code duplication.

rgommers commented:
> It was causing CI failures here so I added it as a commit. Should I make it a separate PR instead?

Ouch. Can you make this a separate PR instead? Then we can merged that straight away; the missing file will be causing test failures for scikit-learn otherwise as soon as they test with a new nightly wheel.

rgommers commented on scipy/optimize/tests/_cython_examples/meson.build:
```diff
@@ -0,0 +1,18 @@
+project('random-build-examples', 'c', 'cpp', 'cython')
+
+py_mod = import('python')
+py3 = py_mod.find_installation(pure: false)
```
Minor: this is more idiomatically written as a one-liner since `py_mod` isn't used for anything:
```meson
py3 = import('python').find_installation(pure: false)
```

This comment applies to the other `meson.build` files too.

rgommers commented on scipy/optimize/tests/_cython_examples/meson.build:
```diff
@@ -0,0 +1,18 @@
+cc = meson.get_compiler('c')
+cy = meson.get_compiler('cython')
+
+# Keep synced with pyproject.toml
```
This should also remain working with older Cython versions, since we're exporting Cython API for other projects to use - and those other projects are not required to upgrade Cython in sync with our version bumps. Hence, I'd just drop this comment; we can leave the lower version bound at 3.0.8 for a long time.

rgommers commented on scipy/optimize/tests/_cython_examples/meson.build:
```diff
@@ -0,0 +1,18 @@
+
+# Keep synced with pyproject.toml
+if not cy.version().version_compare('>=3.0.8')
+  error('tests requires Cython >= 3.0.8')
```
minor: inconsistent indentation between this line and `py3.extension_module` below. Using 4 spaces everywhere is fine, as is 2 (I have no strong preference) - as long as it's consistent.

This comment applies to the other `meson.build` files too.

rgommers commented on scipy/optimize/tests/_cython_examples/meson.build:
```diff
@@ -0,0 +1,18 @@
+project('random-build-examples', 'c', 'cpp', 'cython')
```
You're requiring a C++ compiler here but don't use it then. It may actually be nice to test with C and with C++ codegen separately, since the behavior may differ. If it's done only for one small `.pyx` file per scipy submodule, this seems doable without incurring must test runtime overhead or maintenance cost. WDYT?

thalassemia replied:
That's a good idea. I added this to each test by copying the existing `.pyx` at build time and specifying C++ as the language for Cython transpilation.

rgommers reviewed (commented):
Thanks for tackling this @thalassemia! This looks quite good. I had a few comments on the tests; I didn't look closely at the common setup code, but that looks like a direct copy of what's by now well-tested in numpy so should be all good.

rgommers commented on scipy/linalg/tests/_cython_examples/extending.pyx:
```diff
@@ -0,0 +1,22 @@
+    dgtsv(&n, &nrhs, &a[0], &b[0], &c[0], &x[0], &n, &info)
+
+cpdef float complex complex_dot(float complex[:] cx, float complex[:] cy):
+    """ Take dot prouct of two complex vectors """
```
minor: typo in "product"

tacaswell commented:
I can confirm that sckitlearn complies again with this branch.

thalassemia committed:
TST: Set `CYTHON_CCOMPLEX=0` for extension test

thalassemia committed:
DOC: `CYTHON_CCOMPLEX=0` when using `cython_blas`

thalassemia committed:
TST: Do not mark Cython API tests as slow

[skip circle]
[skip cirrus]

lucascolley commented:
From the linked issue:

> we should be testing with a different version of Cython that we use to build. That is more likely to smoke out issues due to changes in .pxd files.

Is that point covered here?

thalassemia commented:
I added a disclaimer to compile with `-DCYTHON_CCOMPLEX=0` in the documentation for `cython_blas`. I forgot about this flag when writing my `linalg` test extension and briefly panicked when I saw the familiar `cdotu` segfaults in the CI logs. After some tinkering in a local 32-bit Linux Docker image, I confirmed that setting this environment variable fixed the segfaults I observed there. Fortunately, adding this flag also fixed the segfaults in the Windows GHAs, where only the C++ version of my extension was affected (weird...).

On the bright side, this experience taught me a valuable lesson about not prematurely marking a test with `pytest.marks.slow`. I would've been far less confused if I had seen these segfaults in the 32-bit Linux test, which previously skipped these new tests due to the slow marker. As a result, I decided to remove this marker (the new tests only add a couple of seconds each anyway).

thalassemia commented:
> From the linked issue:
> 
> > we should be testing with a different version of Cython that we use to build. That is more likely to smoke out issues due to changes in .pxd files.
> 
> Is that point covered here?

Good catch. Not at the moment, but this would be nice for sure. Is it possible to downgrade Cython after building scipy in a way that does not trigger a Meson rebuild when we invoke `python dev.py test`?

lucascolley commented:
`python dev.py --no-build test` I believe, or just `-n`.

rgommers commented:
> we should be testing with a different version of Cython that we use to build. That is more likely to smoke out issues due to changes in .pxd files.

That is my comment, but I think it is not correct. I believe what I meant is "a different version of SciPy". The failure mode we had before is:
- we change a signature in a pxd file
- scikit-learn compiles against latest scipy, using `cython_blas` in its own Cython code
- scikit-learn then stopped working with older `scipy` versions

Changing Cython versions won't do anything to catch that failure mode. Having a separate CI job where we'd have Cython code built against the dev version of scipy, and in a separate job step tested against an older scipy version, would be nice.

However, this is pretty cumbersome to implement and would require a very different test structure from what is done in this PR now, so I would not attempt to include it here. It'd more be something for a standalone CI job, not the regular test suite probably.

rgommers commented:
> As a result, I decided to remove this marker (the new tests only add a couple of seconds each anyway).

I agree with this. Normally tests this slow require marking as `slow`, however these are basically smoke tests for the most basic functionality provided by the Cython APIs (including "is it installed and cimportable at all"). Compile tests like these are inherently slow, so O(10 sec) added to the default test suite is acceptable I'd say.

More extensive tests, if and when those are added in the future, should be marked as `slow` though.

<details>
<summary>Slowest 40 tests in default (fast) test mode:</summary>

```
========================================================= slowest 40 durations =========================================================
6.85s call     scipy/linalg/tests/test_extending.py::test_cython
5.53s call     scipy/integrate/tests/test_quadpack.py::TestQuad::test_triple_integral_improper[-inf-inf--inf-inf--inf-inf-5.568327996831708]
5.16s call     scipy/stats/tests/test_axis_nan_policy.py::test_axis_nan_policy_fast[mixed-1-omit-cramervonmises_2samp-args44-kwds44-2-2-False-<lambda>]
4.54s call     scipy/stats/tests/test_stats.py::TestWassersteinDistanceND::test_optimization_vs_analytical[300-300]
3.41s call     scipy/optimize/tests/test_optimize.py::test_cobyla_threadsafe
3.35s call     scipy/integrate/tests/test_quadpack.py::TestQuad::test_triple_integral_improper[-inf-1--inf-1--inf-1-4.355111619313947]
3.29s call     scipy/interpolate/tests/test_rgi.py::TestRegularGridInterpolator::test_descending_points_nd[4-<lambda>-quintic] = _wrappers_common.scipy
3.07s call     scipy/optimize/_trustregion_constr/tests/test_report.py::test_gh12922
3.07s call     scipy/stats/tests/test_sensitivity_analysis.py::TestSobolIndices::test_ishigami[scalar]
3.06s call     scipy/integrate/tests/test_quadpack.py::TestQuad::test_triple_integral_improper[-1-inf--1-inf--1-inf-4.355111619313947]
2.94s call     scipy/optimize/tests/test_extending.py::test_cython
2.88s call     scipy/special/tests/test_extending.py::test_cython
2.66s call     scipy/stats/tests/test_distributions.py::TestHypergeom::test_sf_gh18506
2.30s call     scipy/special/tests/test_cython_special.py::test_cython_api[elliprj]
2.23s call     scipy/linalg/tests/test_basic.py::TestLstsq::test_random_complex_exact
2.15s call     scipy/integrate/tests/test_quadpack.py::TestQuad::test_triple_integral_improper[-inf-1--inf-1--inf-0-2.363439379836851]
2.13s call     scipy/integrate/tests/test_quadrature.py::TestCumulativeSimpson::test_cumulative_simpson_against_simpson_with_default_dx
2.09s call     scipy/integrate/tests/test_quadrature.py::TestCumulativeSimpson::test_cumulative_simpson_against_simpson
1.99s call     scipy/integrate/tests/test_quadpack.py::TestQuad::test_triple_integral_improper[-inf-1--inf-0--inf-1-2.363439379836851]
1.97s call     scipy/optimize/tests/test__differential_evolution.py::TestDifferentialEvolutionSolver::test_L1
1.92s call     scipy/integrate/tests/test_quadpack.py::TestQuad::test_triple_integral_improper[-1-inf--1-inf-0-inf-2.363439379836851]
1.88s call     scipy/_lib/tests/test_import_cycles.py::test_public_modules_importable
1.83s call     scipy/stats/tests/test_multivariate.py::TestMultivariateT::test_cdf_against_multivariate_normal[10]
1.81s call     scipy/integrate/tests/test_quadpack.py::TestQuad::test_triple_integral_improper[-inf-0--inf-1--inf-1-2.363439379836851]
1.78s call     scipy/optimize/tests/test__differential_evolution.py::TestDifferentialEvolutionSolver::test_integrality
1.77s call     scipy/optimize/tests/test__differential_evolution.py::TestDifferentialEvolutionSolver::test_L4
1.76s call     scipy/integrate/tests/test_tanhsinh.py::TestTanhSinh::test_accuracy[case97-ref1]
1.75s call     scipy/linalg/tests/test_basic.py::TestLstsq::test_random_exact
1.73s call     scipy/optimize/tests/test_optimize.py::TestOptimizeSimple::test_minimize_callback_copies_array[fmin]
1.72s call     scipy/integrate/tests/test_quadpack.py::TestQuad::test_triple_integral_improper[-1-inf-0-inf--1-inf-2.363439379836851]
1.70s call     scipy/stats/tests/test_continuous_basic.py::test_cont_basic[500-200-ncx2-arg77]
1.65s call     scipy/stats/tests/test_continuous_basic.py::test_cont_basic[500-200-beta-arg4]
1.62s call     scipy/stats/tests/test_resampling.py::TestPower::test_vectorization
1.59s call     scipy/interpolate/tests/test_rgi.py::TestRegularGridInterpolator::test_descending_points_nd[5-<lambda>-slinear]
1.59s call     scipy/integrate/tests/test_quadpack.py::TestQuad::test_triple_integral_improper[0-inf--1-inf--1-inf-2.363439379836851]
1.53s call     scipy/stats/tests/test_fit.py::TestFit::test_basic_fit_mle[vonmises]
1.48s call     scipy/optimize/tests/test_lsq_linear.py::TestTRF::test_sparse_bounds
1.46s call     scipy/stats/tests/test_continuous_basic.py::test_cont_basic[500-200-gengamma-arg31]
1.46s call     scipy/optimize/tests/test__basinhopping.py::TestBasinHopping::test_all_nograd_minimizers
1.45s call     scipy/optimize/tests/test_least_squares.py::TestTRF::test_with_bounds
```

</details>

rgommers reviewed (commented):
Looks great! Just two more comments.

Can you add these to the set of `test` dependencies in `pyproject.toml`, and then run `tools/generate_requirements.py`?
```
Cython
meson
ninja; sys_platform != "emscripten"
```

thalassemia committed:
DOC: ``CYTHON_CCOMPLEX`` define and Cython test requirements

[skip circle]
[skip cirrus]

rgommers reviewed (approved):
LGTM now and CI is green, so let's give it a go. Thanks again @thalassemia!
